File: src/main/java/com/zhihao/admin/module/system/service/MenuService.java
package com.zhihao.admin.module.system.service;

import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.zhihao.admin.common.constant.MenuType;
import com.zhihao.admin.common.constant.SystemConstants;
import com.zhihao.admin.common.exception.BusinessException;
import com.zhihao.admin.common.util.TreeUtils;
import com.zhihao.admin.module.system.dto.MenuRequest;
import com.zhihao.admin.module.system.dto.MenuTreeResponse;
import com.zhihao.admin.module.system.entity.SysMenu;
import com.zhihao.admin.module.system.mapper.SysMenuMapper;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;

@Service
@RequiredArgsConstructor
public class MenuService {

    private final SysMenuMapper menuMapper;

    public List<MenuTreeResponse> tree() {
        List<MenuTreeResponse> nodes = menuMapper.selectList(new LambdaQueryWrapper<>()).stream()
                .map(MenuTreeResponse::from)
                .toList();
        return TreeUtils.build(nodes, MenuTreeResponse::getId, MenuTreeResponse::getParentId,
                MenuTreeResponse::getSort, (parent, child) -> parent.getChildren().add(child),
                SystemConstants.ROOT_PARENT_ID);
    }

    @Transactional(rollbackFor = Exception.class)
    public Long create(MenuRequest request) {
        validateMenuRequest(request);
        SysMenu menu = new SysMenu();
        apply(menu, request);
        menuMapper.insert(menu);
        return menu.getId();
    }

    @Transactional(rollbackFor = Exception.class)
    public void update(Long id, MenuRequest request) {
        validateMenuRequest(request);
        SysMenu menu = requireMenu(id);
        if (id.equals(request.parentId())) {
            throw new BusinessException("menu cannot be its own parent");
        }
        apply(menu, request);
        menuMapper.updateById(menu);
    }

    @Transactional(rollbackFor = Exception.class)
    public void delete(Long id) {
        requireMenu(id);
        Long childCount = menuMapper.selectCount(new LambdaQueryWrapper<SysMenu>().eq(SysMenu::getParentId, id));
        if (childCount > 0) {
            throw new BusinessException("menu has children");
        }
        menuMapper.deleteById(id);
    }

    private SysMenu requireMenu(Long id) {
        SysMenu menu = menuMapper.selectById(id);
        if (menu == null) {
            throw new BusinessException("menu not found");
        }
        return menu;
    }

    private void validateMenuRequest(MenuRequest request) {
        if (MenuType.BUTTON.equals(request.menuType()) && !StringUtils.hasText(request.permission())) {
            throw new BusinessException("button menu requires permission");
        }
        if (MenuType.PAGE.equals(request.menuType()) && !StringUtils.hasText(request.path())) {
            throw new BusinessException("page menu requires path");
        }
    }

    private void apply(SysMenu menu, MenuRequest request) {
        menu.setParentId(request.parentId());
        menu.setMenuName(request.menuName());
        menu.setMenuType(request.menuType());
        menu.setPath(request.path());
        menu.setComponent(request.component());
        menu.setPermission(request.permission());
        menu.setIcon(request.icon());
        menu.setSort(request.sort());
        menu.setVisible(request.visible());
        menu.setStatus(request.status());
    }
}

File: server/src/modules/menu/menu.service.ts
import { ROOT_PARENT_ID } from '../../common/constants'
import { buildTree } from '../../common/tree-utils'
import { BusinessException } from '../../common/errors'
import { prisma } from '../../lib/prisma'
import type { MenuRequest, MenuTreeNode } from './menu.schema'

function toTreeNode(menu: {
  id: bigint
  parentId: bigint
  menuName: string
  menuType: string
  path: string | null
  component: string | null
  permission: string | null
  icon: string | null
  sort: number | null
  visible: number
  status: number
}): MenuTreeNode {
  return {
    id: Number(menu.id),
    parentId: Number(menu.parentId),
    menuName: menu.menuName,
    menuType: menu.menuType,
    path: menu.path ?? undefined,
    component: menu.component ?? undefined,
    permission: menu.permission ?? undefined,
    icon: menu.icon ?? undefined,
    sort: menu.sort ?? undefined,
    visible: menu.visible,
    status: menu.status,
    children: [],
  }
}

function validateMenuRequest(request: MenuRequest) {
  if (request.menuType === 'B' && !request.permission) {
    throw new BusinessException('button menu requires permission')
  }
  if (request.menuType === 'C' && !request.path) {
    throw new BusinessException('page menu requires path')
  }
}

export async function tree(): Promise<MenuTreeNode[]> {
  const menus = await prisma.sysMenu.findMany({ where: { deleted: 0 } })
  const nodes = menus.map(toTreeNode)
  return buildTree(
    nodes,
    (node) => node.id,
    (node) => node.parentId,
    (node) => node.sort,
    (parent, child) => parent.children.push(child),
    ROOT_PARENT_ID,
  )
}

export async function create(request: MenuRequest, operatorId: number | null): Promise<number> {
  validateMenuRequest(request)
  const now = new Date()
  const menu = await prisma.sysMenu.create({
    data: {
      parentId: BigInt(request.parentId),
      menuName: request.menuName,
      menuType: request.menuType,
      path: request.path,
      component: request.component,
      permission: request.permission,
      icon: request.icon,
      sort: request.sort ?? 0,
      visible: request.visible,
      status: request.status,
      createTime: now,
      updateTime: now,
      createBy: operatorId !== null ? BigInt(operatorId) : null,
      updateBy: operatorId !== null ? BigInt(operatorId) : null,
      deleted: 0,
    },
  })
  return Number(menu.id)
}

export async function update(id: number, request: MenuRequest, operatorId: number | null): Promise<void> {
  validateMenuRequest(request)
  await requireMenu(id)
  if (id === request.parentId) {
    throw new BusinessException('menu cannot be its own parent')
  }

  await prisma.sysMenu.update({
    where: { id: BigInt(id) },
    data: {
      parentId: BigInt(request.parentId),
      menuName: request.menuName,
      menuType: request.menuType,
      path: request.path,
      component: request.component,
      permission: request.permission,
      icon: request.icon,
      sort: request.sort ?? 0,
      visible: request.visible,
      status: request.status,
      updateTime: new Date(),
      updateBy: operatorId !== null ? BigInt(operatorId) : null,
    },
  })
}

export async function remove(id: number): Promise<void> {
  await requireMenu(id)
  const childCount = await prisma.sysMenu.count({
    where: { parentId: BigInt(id), deleted: 0 },
  })
  if (childCount > 0) {
    throw new BusinessException('menu has children')
  }
  await prisma.sysMenu.update({
    where: { id: BigInt(id) },
    data: { deleted: 1, updateTime: new Date() },
  })
}

async function requireMenu(id: number) {
  const menu = await prisma.sysMenu.findFirst({
    where: { id: BigInt(id), deleted: 0 },
  })
  if (!menu) {
    throw new BusinessException('menu not found')
  }
  return menu
}
